src/typechecker.erl

```diff
@@ -4227,14 +4343,21 @@ check_guards(Env, Guards) ->
     union_var_binds_symmetrical(Envs, Env).
 
 -spec type_check_function(env(), erl_parse:abstract_form()) -> {env(), constraints:constraints()}.
-type_check_function(Env, {function, _, Name, NArgs, Clauses}) ->
+type_check_function(Env, {function, Anno, Name, NArgs, Clauses} = Func) ->
     ?verbose(Env, "Checking function ~p/~p~n", [Name, NArgs]),
     case maps:find({Name, NArgs}, Env#env.fenv) of
         {ok, FunTy} ->
             NewEnv = Env#env{current_spec = FunTy},
             FunTyNoPos = [ typelib:remove_pos(?assert_type(Ty, type())) || Ty <- FunTy ],
             Arity = clause_arity(hd(Clauses)),
-            check_clauses_fun(NewEnv, expect_fun_type(NewEnv, FunTyNoPos, Arity), Clauses);
+            case expect_fun_type(NewEnv, FunTyNoPos, Arity) of
+                {type_error, NotFunTy} ->
+                    %% This can only happen if `create_fenv/2' creates garbage.
+                    erlang:error({invalid_function_type, NotFunTy});
```

Comment:
Dead code? Shoulnd't we let it crash instead?

Reply:
If we let it crash we'll get a self-check warning. Throwing an error is more explicit and avoids the warning, though it's more verbose.

Reply:
OK